File: microstrain_inertial_examples/src/listener.cpp
/////////////////////////////////////////////////////////////////////////////////////////////////////
//
// Parker-Lord Inertial Device Driver Implementation File
//
// Copyright (c) 2017, Brian Bingham
// Copyright (c) 2020, Parker Hannifin Corp
// This code is licensed under MIT license (see LICENSE file for details)
//
/////////////////////////////////////////////////////////////////////////////////////////////////////
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/imu.hpp"


namespace microstrain {

class Listener : public rclcpp::Node
{
public:
  Listener() : Node("listener")
  {
    // Setup the subscription (lambda for simplicity, but this could also be a member function)
    sub_ = create_subscription<sensor_msgs::msg::Imu>("/imu/data", 10,
      [this](sensor_msgs::msg::Imu::UniquePtr imu)
      {
        RCLCPP_INFO(get_logger(), "Quaternion Orientation:    [%f, %f, %f, %f]", imu->orientation.x, imu->orientation.y, imu->orientation.z,
                imu->orientation.w);
        RCLCPP_INFO(get_logger(), "Angular Velocity:          [%f, %f, %f]", imu->angular_velocity.x, imu->angular_velocity.y,
                imu->angular_velocity.z);
        RCLCPP_INFO(get_logger(), "Linear Acceleration:       [%f, %f, %f]", imu->linear_acceleration.x, imu->linear_acceleration.y,
                imu->linear_acceleration.z);

        // add code here to handle incoming IMU data
      }
    );
  }

private:
  rclcpp::Subscription<sensor_msgs::msg::Imu>::SharedPtr sub_;
};

};  // namespace microstrain

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<microstrain::Listener>());
  rclcpp::shutdown();
  return 0;
}

File: microstrain_inertial_examples/scripts/listener.py
#! /usr/bin/env python3
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Imu

class Listener(Node):
    def __init__(self):
        super().__init__("listener")
        self.subscription = self.create_subscription(Imu, '/imu/data', self.imuDataCallback, 10)

    def imuDataCallback(self, imu):
        self.get_logger().info("Quaternion Orientation:  [%f, %f, %f, %f]" % (imu.orientation.x, imu.orientation.y, imu.orientation.z, imu.orientation.w));
        self.get_logger().info("Angular Velocity:        [%f, %f, %f]" % (imu.angular_velocity.x, imu.angular_velocity.y, imu.angular_velocity.z));
        # add code here to handle incoming IMU data
    
if __name__ == '__main__':
    rclpy.init(args=None)

    listener = Listener()

    rclpy.spin(listener)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    listener.destroy_node()
    rclpy.shutdown()
